404 errors don't get rendered correctly by the api-lambda

404s don't get rendered correctly for API routes. This is supposed to be handled by the _error.js, but that returns 502 for me with a CloudWatch message from cloudfront: "ERROR Validation error: The Lambda function returned an invalid json output, parsed json should be an object type." 

^ Actually to get to that point took a workaround. You would normally see a 503 for 404'ed /api routes (in serverless trace mode) with a "Cannot find module 'next/dist/next-server/server/render" CloudWatch error message. This happens because buildApiLambda() is missing _error.js in its page list prior to calling nodeFileTrace. This causes that page's dependencies to not get bundled correctly. A simple `apiPages.push(path.join(this.serverlessDir, "pages/_error.js"));` there can fix it there to get to the 502 final boss that I'm currently working through.

Ideally, the api-lambda should return the static 404.html for all not-found routes.

--- a/packages/lambda-at-edge/src/api-handler.ts
+++ b/packages/lambda-at-edge/src/api-handler.ts
@@ -8,12 +8,12 @@ const normaliseUri = (uri: string): string => (uri === "/" ? "/index" : uri);
 
 const router = (
   manifest: OriginRequestApiHandlerManifest
-): ((path: string) => string) => {
+): ((path: string) => string | null) => {
   const {
     apis: { dynamic, nonDynamic }
   } = manifest;
 
-  return (path: string): string => {
+  return (path: string): string | null => {
     if (nonDynamic[path]) {
       return nonDynamic[path];
     }
@@ -29,7 +29,7 @@ const router = (
       }
     }
 
-    return "pages/_error.js";
+    return null;
   };
 };
 
@@ -40,6 +40,12 @@ export const handler = async (
   const uri = normaliseUri(request.uri);
 
   const pagePath = router(manifest)(uri);
+
+  if (!pagePath) {
+    return {
+      status: "404"
+    };
+  }
 
   // eslint-disable-next-line
   const page = require(`./${pagePath}`);

--- a/packages/lambda-at-edge/src/build.ts
+++ b/packages/lambda-at-edge/src/build.ts
@@ -251,10 +251,6 @@ class Builder {
         join(this.serverlessDir, "pages/api"),
         join(this.outputDir, API_LAMBDA_CODE_DIR, "pages/api")
       ),
-      fse.copy(
-        join(this.serverlessDir, "pages/_error.js"),
-        join(this.outputDir, API_LAMBDA_CODE_DIR, "pages/_error.js")
-      ),
       fse.writeJson(
         join(this.outputDir, API_LAMBDA_CODE_DIR, "manifest.json"),
         apiBuildManifest

--- a/packages/lambda-at-edge/tests/api-handler/api-handler.test.ts
+++ b/packages/lambda-at-edge/tests/api-handler/api-handler.test.ts
@@ -41,4 +41,22 @@ describe("API lambda handler", () => {
     expect(decodedBody).toEqual("pages/api/getCustomers");
     expect(response.status).toEqual(200);
   });
+
+  it("returns 404 for not-found api routes", async () => {
+    const event = createCloudFrontEvent({
+      uri: "/foo/bar",
+      host: "mydistribution.cloudfront.net",
+      origin: {
+        s3: {
+          domainName: "my-bucket.amazonaws.com"
+        }
+      }
+    });
+
+    mockPageRequire("pages/api/getCustomers.js");
+
+    const response = (await handler(event)) as CloudFrontResponseResult;
+
+    expect(response.status).toEqual("404");
+  });
 });

--- a/packages/lambda-at-edge/tests/build/build.test.ts
+++ b/packages/lambda-at-edge/tests/build/build.test.ts
@@ -222,7 +222,7 @@ describe("Builder Tests", () => {
         "node_modules",
         "pages"
       ]);
-      expect(pages).toEqual(["_error.js", "api"]);
+      expect(pages).toEqual(["api"]);
     });
   });
 });

